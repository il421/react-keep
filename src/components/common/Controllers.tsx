import React from "react";
import { AlignItems, JustifyContent } from "../../common";
import { FlexBox, IconButton } from "../ui-components";
import "../../styles/components/common/_controllers.scss";
import { NoteType } from "../notes";

interface Controllers extends Partial<Pick<HTMLElement, "className">> {
  isMobile?: boolean;
  openDialog: (type: NoteType) => void;
}

export const Controllers: React.FunctionComponent<Controllers> = ({
  className,
  isMobile = false,
  openDialog,
}) => {
  const handleOnClick = (type: NoteType) => () => openDialog(type);

  return (
    <div
      className={`controllers  ${
        isMobile && "controllers--mobile"
      } ${className}`}
    >
      <FlexBox
        vertical={isMobile}
        justifyContent={JustifyContent.spaceBetween}
        alignItems={AlignItems.center}
      >
        <IconButton
          className="controllers__text"
          onButtonClick={handleOnClick(NoteType.text)}
          icon="align-left"
          size="2x"
        />

        <IconButton
          className="controllers__list"
          onButtonClick={handleOnClick(NoteType.list)}
          icon="check-square"
          size="2x"
        />

        <IconButton
          className="controllers__image"
          onButtonClick={handleOnClick(NoteType.image)}
          icon="image"
          size="2x"
        />
      </FlexBox>
    </div>
  );
};
