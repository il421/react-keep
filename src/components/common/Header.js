import React from 'react';
import { connect } from 'react-redux';

import { startLogout } from '../../actions/auth';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import Search from './Search';

export const Header = ({ startLogout, auth, showSidebar }) => {

  return (
    <header className="header" onClick={(evt) => evt.stopPropagation()}>
      <div className="content-container">
        <div className="header__content">
          <div className="header__sidebar">

            <button className="button button--sidenav" onClick={() => showSidebar()}>
              <FontAwesomeIcon icon="bars" size="2x" />
            </button>

          </div>
          <div>
            <h1 className="header__title">Keep Me</h1>
          </div>
          <div className="header__search">
            <Search />
          </div>
          <div className="header__user-box user-box">
            <div className="user-box__photo">
              <img src={ auth.url } width={ 50 } height={ 50 } />
            </div>
            <div className="user-box__name show-for-mobile">{ auth.name }</div>
            <button className="user-box__logout button button--sidenav" onClick={ startLogout }>
              <FontAwesomeIcon icon="sign-out-alt" size="lg" />
            </button>
          </div>
        </div>
      </div>
    </header>
  );
};


const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
    url: state.url,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
