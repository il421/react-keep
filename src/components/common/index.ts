export * from "./Controllers";
export * from "./DashboardPage";
export * from "./Header";
export * from "./NotFoundPage";
export * from "./Search";
export * from "./SideBar";
