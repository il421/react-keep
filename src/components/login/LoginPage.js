import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../../actions/auth';

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout login">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Keep Me</h1>
      <p>Design by Ilya Suglobov</p>
      <button className="button login__btn" onClick={ startLogin }>Login via Google</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
