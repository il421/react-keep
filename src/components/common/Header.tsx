import React from "react";
import { connect } from "react-redux";
import { useHistory } from "react-router-dom";
import { stringify } from "query-string";
import { AuthStoreState, Store } from "../../store/store.types";
import { QueryKeys, RouteActions } from "../../routers/Routing";
import { startLogout } from "../../actions/auth";
import Search from "./Search";
import { Controllers } from "./Controllers";
import {
  IconButton,
  UserPhoto,
  ContentContainer,
  LinkButton
} from "../ui-components";
import "../../styles/components/header/_header.scss";
import "../../styles/components/header/_user-box.scss";
import "../../styles/components/common/_visibility.scss";
import { ThunkDispatch } from "redux-thunk";

interface HeaderProps {
  showSidebar: (value: boolean) => void;
}

interface StateProps {
  auth: AuthStoreState;
}

interface DispatchProps {
  startLogout: () => void;
}

type Props = HeaderProps & DispatchProps & StateProps;

const Header: React.FunctionComponent<Props> = ({
  startLogout,
  auth,
  showSidebar
}) => {
  const history = useHistory();

  const onClickHandler = (key: keyof typeof QueryKeys) => {
    const query = stringify({
      [key]: RouteActions.edit
    });
    history.push(`${history.location.pathname}?${query}`);
  };

  return (
    <header className="header" onClick={evt => evt.stopPropagation()}>
      <ContentContainer>
        <div className="header__content">
          <IconButton
            className="header__bar"
            icon="bars"
            size="2x"
            onClick={() => showSidebar(true)}
          />

          <h1 className="header__title">Keep Me</h1>

          <Search wrapperClass="header__search" />

          <Controllers className="hide-for-mobile" />

          <div className="header__user-box user-box">
            {auth.name && auth.url && (
              <UserPhoto
                className="user-box__photo"
                src={auth.url}
                height={50}
                width={50}
              />
            )}
            <LinkButton
              text={auth.name ?? "Update profile"}
              type="button"
              className="user-box__link"
              onClick={() => onClickHandler(QueryKeys.user)}
            />
            <IconButton
              className="user-box__logout"
              icon="sign-out-alt"
              size="lg"
              onClick={startLogout}
            />
          </div>
        </div>
      </ContentContainer>
    </header>
  );
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state: Store): StateProps => {
  return {
    auth: state.auth
  };
};

export default connect<StateProps, DispatchProps, HeaderProps, Store>(
  mapStateToProps,
  mapDispatchToProps
)(Header);
