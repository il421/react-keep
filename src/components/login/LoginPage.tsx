import React from "react";
import { connect } from "react-redux";
import { ValidationErrors } from "final-form";
import { ToastContainer } from "react-toastify";
import { startLogin, startSignUp } from "../../actions/auth";
import { Store } from "../../store/store.types";
import { nameOf, Errors, emailRegEx, Placeholders } from "../../common";
import "../../styles/components/login/_login-page.scss";
import { BaseForm, CheckboxInputField, TextInputField } from "../form";
import { ConfirmButton } from "../ui-components";
import { ThunkDispatch } from "redux-thunk";

export interface LoginFormValues {
  email: string;
  password: string;
  isNew: boolean;
}

interface StateProps {
  loading: boolean;
}

interface DispatchProps {
  startLogin: (email: string, password: string) => void;
  startSignUp: (email: string, password: string) => void;
}

export type Props = StateProps & DispatchProps;
export const getValidationErrors = (
  values: LoginFormValues
): ValidationErrors => {
  const errors: ValidationErrors = {};
  const { email, password } = values;

  if (!password) {
    errors.password = Errors.required;
  } else if (password.length < 6) {
    errors.password = Errors.password;
  }

  if (!email) {
    errors.email = Errors.required;
  } else if (!emailRegEx.test(email)) {
    errors.email = Errors.email;
  }

  return errors;
};

export class LoginPage extends React.PureComponent<Props> {
  private initialValues: LoginFormValues = {
    email: "",
    password: "",
    isNew: false,
  };

  private nameOf = nameOf<LoginFormValues>();

  private getButtons = (
    isDisable: boolean,
    isSubmitting: boolean,
    values: LoginFormValues
  ) => {
    return (
      <ConfirmButton
        className="login-button"
        text={values.isNew ? "Sign Up" : "Log In"}
        loading={this.props.loading}
        disabled={isDisable}
      />
    );
  };

  private submitLoginForm = (values: LoginFormValues): void => {
    if (values.isNew) {
      this.props.startSignUp(values.email, values.password);
    } else {
      this.props.startLogin(values.email, values.password);
    }
  };

  render() {
    return (
      <div className="login-page">
        <div className="login-page__login-box login-box">
          <div className="login-box__desc">
            <h1>Keep Me</h1>
            <p>Design by Ilya Suglobov</p>
          </div>

          <BaseForm<LoginFormValues>
            initialValues={this.initialValues}
            classNames={{
              form: "login-box__form login-form",
            }}
            onSubmit={this.submitLoginForm}
            validate={getValidationErrors}
            getButtons={this.getButtons}
          >
            <TextInputField
              name={this.nameOf("email")}
              type="text"
              placeholder={Placeholders.email}
              className="login-form__field"
            />

            <TextInputField
              name={this.nameOf("password")}
              type="password"
              placeholder={Placeholders.password}
              className="login-form__field"
            />

            <div className="login-form__new-user-field new-user-field">
              <CheckboxInputField
                id="is-new-user"
                name={this.nameOf("isNew")}
                className="new-user-field__checkbox"
              />
              <div>I want to sign up as a new user</div>
            </div>
          </BaseForm>
        </div>
        <ToastContainer autoClose={3000} />
      </div>
    );
  }
}

const mapStateToProps = (state: Store): StateProps => {
  return {
    loading: state.auth.loading,
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  startLogin: (email: string, password: string) =>
    dispatch(startLogin(email, password)),
  startSignUp: (email: string, password: string) =>
    dispatch(startSignUp(email, password)),
});

export default connect<StateProps, DispatchProps, {}, Store>(
  mapStateToProps,
  mapDispatchToProps
)(LoginPage);
