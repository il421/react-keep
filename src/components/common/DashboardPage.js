import React, { useState } from 'react';

import NoteSelection from '../note/NoteSelection';
import NotesList from '../note/NotesList';
import Header from './Header';
import SideBar from './SideBar';


const DashboardPage = () => {
  let [showBar, toggleSidebarDisplay] = useState(false);

  const showSidebar = () => {
    toggleSidebarDisplay(showBar = true);
  };

  const hideSidebar = () => {
    toggleSidebarDisplay(showBar = false);
  };

  return (
    <div className="dashboard">
      <Header showSidebar={ showSidebar } />
      <NoteSelection className="content-container"/>
      <NotesList/>

      <SideBar showBar={ showBar } hideSidebar={ hideSidebar } />
    </div>
  );
};
export default DashboardPage;
