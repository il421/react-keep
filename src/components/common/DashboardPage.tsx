import React from "react";
import { ToastContainer } from "react-toastify";
import { History, Path } from "history";
import { JustifyContent } from "../../common";
import { Header } from "./Header";
import { SideBar } from "./SideBar";
import { Controllers } from "./Controllers";
import UserFormModal from "../login/UserFormModal";
import "../../styles/components/common/_dashboard.scss";
import "../../styles/components/notes/_note-modal.scss";
import { QueryKeys } from "../../routers/Routing";
import { stringify } from "query-string";
import {
  NoteType,
  ImageNoteFormModal,
  TextNoteFormModal,
  ListNoteFormModal,
} from "../notes";
import { FlexBox } from "../ui-components";
import { NotesList } from "../notes/NotesList";

export interface DashboardPageProps {
  history: History;
}

export const onNoteSelected = (options: {
  type: NoteType;
  id: string;
  pathname: string;
  push: (path: Path) => void;
}) => {
  let key: keyof typeof QueryKeys;
  const { type, id, pathname, push } = options;
  switch (type) {
    case NoteType.text:
      key = QueryKeys.text;
      break;
    case NoteType.list:
      key = QueryKeys.list;
      break;
    default:
      key = QueryKeys.image;
  }

  const query = stringify({
    [key]: id,
  });
  push(`${pathname}?${query}`);
};

export const DashboardPage: React.FunctionComponent<DashboardPageProps> = ({
  history,
}): JSX.Element => {
  return (
    <FlexBox
      vertical={true}
      justifyContent={JustifyContent.spaceBetween}
      className="dashboard"
    >
      <Header history={history} />
      <NotesList
        onNoteSelected={(type: NoteType, id: string) =>
          onNoteSelected({
            type,
            id,
            pathname: history.location.pathname,
            push: history.push,
          })
        }
      />

      <SideBar />
      <TextNoteFormModal history={history} />
      <ListNoteFormModal history={history} />
      <ImageNoteFormModal history={history} />
      <UserFormModal />
      <Controllers
        isMobile
        openDialog={(query) =>
          history.push(`${history.location.pathname}?${query}`)
        }
      />

      <ToastContainer autoClose={2000} />
    </FlexBox>
  );
};
