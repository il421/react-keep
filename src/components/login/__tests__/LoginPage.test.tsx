import { mount, ReactWrapper } from "enzyme";
import { ValidationErrors } from "final-form";
import React from "react";

import { Errors } from "../../../common";
import {
  triggerCheckboxChange,
  triggerInputChange
} from "../../../common/testUtils";
import { user } from "../../../testData/users";
import { Props, LoginPage, getValidationErrors } from "../LoginPage";

let wrapper: ReactWrapper | undefined, props: Props;

beforeEach(() => {
  props = {
    startLogin: jest.fn(),
    startSignUp: jest.fn(),
    loading: false
  };
  wrapper = mount(<LoginPage {...props} />);
});

afterEach(() => {
  if (wrapper) {
    wrapper.unmount();
    wrapper = undefined;
  }
});

describe("Rendering", () => {
  test("should render LoginPage correctly", () => {
    if (wrapper) {
      expect(wrapper.debug()).toMatchSnapshot();
    }
  });
});

describe("Submitting", () => {
  test("should logIn with  correct data", () => {
    if (wrapper) {
      triggerInputChange(wrapper, { name: "email" }, user.email);
      triggerInputChange(wrapper, { name: "password" }, user.password);

      wrapper.find("form").simulate("focus").simulate("submit");
      expect(props.startLogin).toHaveBeenLastCalledWith(
        user.email,
        user.password
      );
      expect(props.startSignUp).not.toBeCalled();
    }
  });

  test("should submit logIn correct data correctly", () => {
    if (wrapper) {
      triggerInputChange(wrapper, { name: "email" }, user.email);
      triggerInputChange(wrapper, { name: "password" }, user.password);
      triggerCheckboxChange(wrapper, { name: "isNew" }, true);

      wrapper.find("form").simulate("focus").simulate("submit");
      expect(props.startSignUp).toHaveBeenLastCalledWith(
        user.email,
        user.password
      );
      expect(props.startLogin).not.toBeCalled();
    }
  });

  test("should not submit with invalid email correctly", () => {
    if (wrapper) {
      triggerInputChange(wrapper, { name: "email" }, "email");
      triggerInputChange(wrapper, { name: "password" }, user.password);
      triggerCheckboxChange(wrapper, { name: "isNew" }, true);

      wrapper.find("form").simulate("focus").simulate("submit");
      expect(props.startSignUp).not.toBeCalled();
      expect(props.startLogin).not.toBeCalled();
    }
  });
});

describe("Validation", () => {
  test("should return no errors", () => {
    const noErrors: ValidationErrors = getValidationErrors({
      email: user.email,
      password: user.password,
      isNew: false
    });
    expect(Object.keys(noErrors).length).toBe(0);
  });

  test("should return email required", () => {
    const emailIsRequiredError: ValidationErrors = getValidationErrors({
      email: "",
      password: user.password,
      isNew: false
    });
    expect(Object.keys(emailIsRequiredError).length).toBe(1);
    expect(emailIsRequiredError.email).toBe(Errors.required);
  });

  test("should return email is not valid", () => {
    const emailIsRequiredError: ValidationErrors = getValidationErrors({
      email: "il.com",
      password: user.password,
      isNew: false
    });
    expect(Object.keys(emailIsRequiredError).length).toBe(1);
    expect(emailIsRequiredError.email).toBe(Errors.email);
  });

  test("should return password required", () => {
    const emailIsRequiredError: ValidationErrors = getValidationErrors({
      email: user.email,
      password: "",
      isNew: false
    });
    expect(Object.keys(emailIsRequiredError).length).toBe(1);
    expect(emailIsRequiredError.password).toBe(Errors.required);
  });

  test("should return password to short", () => {
    const emailIsRequiredError: ValidationErrors = getValidationErrors({
      email: user.email,
      password: "123",
      isNew: false
    });
    expect(Object.keys(emailIsRequiredError).length).toBe(1);
    expect(emailIsRequiredError.password).toBe(Errors.password);
  });
});
