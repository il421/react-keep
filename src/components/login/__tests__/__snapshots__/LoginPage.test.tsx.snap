// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should render LoginPage correctly 1`] = `
"<LoginPage startLogin={[Function: mockConstructor]} startSignUp={[Function: mockConstructor]} loading={false}>
  <div className=\\"login-page\\">
    <div className=\\"login-page__login-box login-box\\">
      <div className=\\"login-box__desc\\">
        <h1>
          Keep Me
        </h1>
        <p>
          Design by Ilya Suglobov
        </p>
      </div>
      <BaseForm initialValues={{...}} formClassName=\\"login-box__form login-form\\" onSubmit={[Function]} validate={[Function: getValidationErrors]} getFormActions={[Function]}>
        <ReactFinalForm initialValues={{...}} onSubmit={[Function]} validate={[Function: getValidationErrors]} mutators={{...}}>
          <form onSubmit={[Function: handleSubmit]} className=\\"login-box__form login-form\\" autoComplete=\\"off\\">
            <TextInputField autoFocus={true} name=\\"email\\" type=\\"text\\" placeholder=\\"Email\\" className=\\"login-form__field\\">
              <Field autoFocus={true} name=\\"email\\" type=\\"text\\" placeholder=\\"Email\\" className=\\"login-form__field\\" component={[Function: FieldAdapter]} parse={[Function: parse]}>
                <FieldAdapter input={{...}} meta={{...}} autoFocus={true} placeholder=\\"Email\\" className=\\"login-form__field\\" render={[undefined]}>
                  <div className=\\"login-form__field\\">
                    <input name=\\"email\\" value=\\"\\" checked={[undefined]} onBlur={[Function]} onChange={[Function]} onFocus={[Function]} type=\\"text\\" placeholder=\\"Email\\" render={[undefined]} onInput={[Function: handleOnChange]} onPaste={[undefined]} autoComplete=\\"new-password\\" autoFocus={true} />
                  </div>
                </FieldAdapter>
              </Field>
            </TextInputField>
            <TextInputField name=\\"password\\" type=\\"password\\" placeholder=\\"Password\\" className=\\"login-form__field\\">
              <Field name=\\"password\\" type=\\"password\\" placeholder=\\"Password\\" className=\\"login-form__field\\" component={[Function: FieldAdapter]} parse={[Function: parse]}>
                <FieldAdapter input={{...}} meta={{...}} placeholder=\\"Password\\" className=\\"login-form__field\\" render={[undefined]}>
                  <div className=\\"login-form__field\\">
                    <input name=\\"password\\" value=\\"\\" checked={[undefined]} onBlur={[Function]} onChange={[Function]} onFocus={[Function]} type=\\"password\\" placeholder=\\"Password\\" render={[undefined]} onInput={[Function: handleOnChange]} onPaste={[undefined]} autoComplete=\\"new-password\\" autoFocus={false} />
                  </div>
                </FieldAdapter>
              </Field>
            </TextInputField>
            <div className=\\"login-form__new-user-field new-user-field\\">
              <CheckboxInputField id=\\"is-new-user\\" name=\\"isNew\\" className=\\"new-user-field__checkbox\\">
                <Field id=\\"is-new-user\\" name=\\"isNew\\" className=\\"new-user-field__checkbox\\" type=\\"checkbox\\" component={[Function: FieldAdapter]}>
                  <FieldAdapter input={{...}} meta={{...}} id=\\"is-new-user\\" className=\\"new-user-field__checkbox\\" render={[undefined]}>
                    <div className=\\"new-user-field__checkbox\\">
                      <input id=\\"is-new-user\\" name=\\"isNew\\" value={[undefined]} checked={false} onBlur={[Function]} onChange={[Function]} onFocus={[Function]} type=\\"checkbox\\" />
                      <label htmlFor=\\"is-new-user\\" />
                    </div>
                  </FieldAdapter>
                </Field>
              </CheckboxInputField>
              <div>
                I want to sign up as a new user
              </div>
            </div>
            <div>
              <ConfirmButton className=\\"login-button\\" text=\\"Log In\\" loading={false} disabled={true}>
                <FlexBox className={[undefined]} justifyContent=\\"center\\">
                  <div style={{...}} className={[undefined]}>
                    <button id={[undefined]} className=\\"login-button\\" type=\\"submit\\" disabled={true} onClick={[undefined]}>
                      Log In
                    </button>
                  </div>
                </FlexBox>
              </ConfirmButton>
            </div>
          </form>
        </ReactFinalForm>
      </BaseForm>
    </div>
    <ToastContainer autoClose={3000} position=\\"top-right\\" transition={[Function: Animation]} rtl={false} hideProgressBar={false} closeButton={{...}} pauseOnHover={true} pauseOnFocusLoss={true} closeOnClick={true} newestOnTop={false} draggable={true} draggablePercent={80} className={{...}} style={{...}} toastClassName={{...}} bodyClassName={{...}} progressClassName={{...}} progressStyle={{...}} role=\\"alert\\">
      <div className=\\"Toastify\\" />
    </ToastContainer>
  </div>
</LoginPage>"
`;
